### PR TITLE
kube-apiserver/corerest: normalize service IP range plumbing

### PR DESCRIPTION
**Pull request by sttts**
The service IP range logic in the core REST storage provider is pretty complicated. This PR cleans up that code area by trying to move that complexity into a central function, not leaking into the main storage setup code.

This is pretty much just a code move.

Extracted from https://github.com/kubernetes/kubernetes/pull/119042. The split into generic-vs-non-generic is towards https://github.com/kubernetes/enhancements/pull/4052.

/kind cleanup

```release-note
NONE
```

### CODE REVIEW

File: pkg/registry/core/rest/storage_core.go [hunk 1]
```diff
@@ -92,16 +86,28 @@ type LegacyRESTStorageProvider struct {
...
+
+	// ServiceIPRange is used to build cluster IPs for discovery.
+	ServiceIPRange net.IPNet
+
+	// allocates ips for secondary service cidr in dual  stack clusters
+	SecondaryServiceIPRange net.IPNet
+	ServiceNodePortRange    utilnet.PortRange
+}
+
+type ServiceIPRangeRegistries struct {
```
Comment (sttts) [inline pkg/registry/core/rest/storage_core.go:104]: this will become private in follow-up PR.

**Comment by sttts** [comment]
```
Kubernetes e2e suite: [It] [sig-network] HostPort validates that there is no conflict between pods with same hostPort but different hostIP and protocol [LinuxOnly] [Conformance] expand_less | 40s
-- | --
{ failed [FAILED] wait for pod pod3 timeout, err:Told to stop trying after 2.016s. The phase of Pod pod3 is Failed which is unexpected. In [It] at: test/e2e/network/hostport.go:219 @ 07/05/23 08:07:29.322 }
```

looks suspicous that I broke something. Looking.

**Comment by aojea** [comment]
> ```
> Kubernetes e2e suite: [It] [sig-network] HostPort validates that there is no conflict between pods with same hostPort but different hostIP and protocol [LinuxOnly] [Conformance] expand_less | 40s
> -- | --
> { failed [FAILED] wait for pod pod3 timeout, err:Told to stop trying after 2.016s. The phase of Pod pod3 is Failed which is unexpected. In [It] at: test/e2e/network/hostport.go:219 @ 07/05/23 08:07:29.322 }
> ```
> 
> looks suspicous that I broke something. Looking.

unrelated, failed with

> Jul  5 08:07:29.352: INFO: At 2023-07-05 08:07:27 +0000 UTC - event for pod3: {kubelet kind-worker2} OutOfpods: Node didn't have enough resource: pods, requested: 1, used: 110, capacity: 110

I saw it failing in other PRs too

File: pkg/registry/core/rest/storage_core.go [hunk 2]
```diff
@@ -300,7 +342,20 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(apiResourceConfigSource
 		return kubernetesservice.RangeRegistries{}, genericapiserver.APIGroupInfo{}, err
 	}
 
-	storage := map[string]rest.Storage{}
+	storage := apiGroupInfo.VersionedResourcesStorageMap["v1"]
+	if storage == nil {
+		storage = map[string]rest.Storage{}
+	}
```
Comment (aojea) [inline pkg/registry/core/rest/storage_core.go:252]: I do not understand well, you initilaze storage here and then check if is nil?

**Comment by aojea** [comment]
LGTM for the allocators part, @wojtek-t for the storage constructors part, I'm not well versed on that part

**Comment by sttts** [comment]
> for the storage constructors part, I'm not well versed on that part

It's just a map of REST Storage objects, indexed by resource names. As we have two funcs now, in theory the first could return an empty `storage` for `v1`. Hence, we have to check for emptiness. We don't want to replace the one of the first func, just extend. No magic there.

File: pkg/registry/core/rest/storage_core.go [hunk 3]
```diff
@@ -325,13 +283,32 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(apiResourceConfigSource
...
+		if storage[resource+"/token"] != nil {
+			storage[resource+"/token"].Destroy()
+		}
+
+		storage[resource] = serviceAccountStorage
+		if serviceAccountStorage.Token != nil {
+			storage[resource+"/token"] = serviceAccountStorage.Token
+		}
+	}
+
```
Comment (sttts) [inline pkg/registry/core/rest/storage_core.go:311]: this block is the only non-move bit in this PR. The generic func create a service account storage without pod support. Here we overwrite it.

Am open for other patterns, didn't find one that I like.

Comment (wojtek-t) [reply pkg/registry/core/rest/storage_core.go:311]: I don't have a better one on my mind either...

File: pkg/registry/core/rest/storage_core.go [hunk 4]
```diff
@@ -110,48 +109,103 @@ func (c LegacyRESTStorageProvider) NewLegacyRESTStorage(apiResourceConfigSource
...
+	namespaceStorage, namespaceStatusStorage, namespaceFinalizeStorage, err := namespacestore.NewREST(restOptionsGetter)
 	if err != nil {
-		return kubernetesservice.RangeRegistries{}, genericapiserver.APIGroupInfo{}, err
+		return genericapiserver.APIGroupInfo{}, err
 	}
 
-	resourceQuotaStorage, resourceQuotaStatusStorage, err := resourcequotastore.NewREST(restOptionsGetter)
+	var serviceAccountStorage *serviceaccountstore.REST
+	if c.ServiceAccountIssuer != nil {
+		serviceAccountStorage, err = serviceaccountstore.NewREST(restOptionsGetter, c.ServiceAccountIssuer, c.APIAudiences, c.ServiceAccountMaxExpiration, nil, secretStorage.Store, c.ExtendExpiration)
```
[the remark this answers is not in the document]
Comment (liggitt) [reply pkg/registry/core/rest/storage_core.go:138]: @sttts just got here as part of the review of https://github.com/kubernetes/kubernetes/pull/120780/files#r1363046940

This change is *really* hard to reason about... the compensation path @wojtek-t noted below is in a separate function ... what gets returned from GenericLegacyRESTStorageProvider#NewRESTStorage here is broken / will panic if a pod-bound token is requested.

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/registry/core/rest/storage_core.go:138]: A panic is definitely unfortunate. Rejecting that token is expected, aka no support for pod bound tokens (it's the generic variant after all, without pod APIs).

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/registry/core/rest/storage_core.go:138]: Passing a getter that always returns NotFound, would do, wouldn't it?

[the remark this answers is not in the document]
Comment (sttts) [reply pkg/registry/core/rest/storage_core.go:138]: https://github.com/kubernetes/kubernetes/pull/121372

**Pull request merged**